### PR TITLE
Arbitrary dim for slice

### PR DESCRIPTION
**Pull request by yanchen036**
Add arbitrary dim support for slice op(https://github.com/tensorflow/tensorflow/issues/8873)
There is still a compiling problem. When build //tensorflow/core/kernel:slice_op_test，I got the following errors:
```
ERROR: /home/parallels/Desktop/github/tensorflow/tensorflow/core/kernels/BUILD:1153:1: Linking of rule '//tensorflow/core/kernels:slice_op_test' failed: gcc failed: error executing command /usr/bin/gcc -o bazel-out/local-opt/bin/tensorflow/core/kernels/slice_op_test '-Wl,-rpath,$ORIGIN/../../../_solib_k8/' -Lbazel-out/local-opt/bin/_solib_k8 -pthread '-fuse-ld=gold' -Wl,-no-as-needed ... (remaining 6 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
bazel-out/local-opt/bin/tensorflow/core/kernels/_objs/slice_op/tensorflow/core/kernels/slice_op.o:slice_op.cc:function tensorflow::SliceOp<Eigen::ThreadPoolDevice, Eigen::QInt8>::Compute(tensorflow::OpKernelContext*): error: undefined reference to 'void tensorflow::internal::SliceSimple<Eigen::ThreadPoolDevice, Eigen::QInt8>(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, tensorflow::gtl::ArraySlice<long long> const&)'
bazel-out/local-opt/bin/tensorflow/core/kernels/_objs/slice_op/tensorflow/core/kernels/slice_op.o:slice_op.cc:function tensorflow::SliceOp<Eigen::ThreadPoolDevice, Eigen::QUInt8>::Compute(tensorflow::OpKernelContext*): error: undefined reference to 'void tensorflow::internal::SliceSimple<Eigen::ThreadPoolDevice, Eigen::QUInt8>(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, tensorflow::gtl::ArraySlice<long long> const&)'
bazel-out/local-opt/bin/tensorflow/core/kernels/_objs/slice_op/tensorflow/core/kernels/slice_op.o:slice_op.cc:function tensorflow::SliceOp<Eigen::ThreadPoolDevice, Eigen::QInt32>::Compute(tensorflow::OpKernelContext*): error: undefined reference to 'void tensorflow::internal::SliceSimple<Eigen::ThreadPoolDevice, Eigen::QInt32>(Eigen::ThreadPoolDevice const&, tensorflow::Tensor*, tensorflow::Tensor const&, tensorflow::gtl::ArraySlice<long long> const&)'
```
It looks like there are some types which haven't be instantiated. However, I can't find that where are these type used. And simply add these instantiation is not a good idea. Could you give some help? @girving

### CODE REVIEW

**Comment by girving** [comment]
I think the `SliceSimple` problem is that you are declaring instantiations for `Slice` in a file that defines `SliceSimple`.  But the place that uses `Slice` doesn't know that you've already instantiated it, so it instantiates it again (wasting code).  And the file that declares `SliceSimple` doesn't know that `SliceSimple` needs to be used outside of that file, so it does a locally visible instantiation (or perhaps inlines everything.

Bottom line: You should declare instantiations of stuff you define in that file, not other stuff.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Why are we merging the `cpu_impl` files?  As I understand it, they take a long time so that's why we split them.

File: tensorflow/core/kernels/slice_op_cpu_impl_6.cc [hunk 1]
```diff
@@ -1,18 +0,0 @@
-/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
Comment (drpngx) [inline tensorflow/core/kernels/slice_op_cpu_impl_6.cc]: These files are here to minimize the compile time.

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/slice_op_cpu_impl_6.cc (hunk 1 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op_cpu_impl_6.cc]: I see, but I delete the template parameter "NDIM", how could I bring these files back? There is similar implementation in transpose_op https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/transpose_functor_cpu.cc#L55

**Comment by yanchen036** [comment]
fixed compiling problem and pass python unit test

**Comment by yanchen036** [comment]
@girving could you test this code please?

**Comment by girving** [comment]
Unfortunately I no longer work at Google, and am currently on vacation. 
@drpngx: Can you take over?

**Comment by drpngx** [comment]
Jenkins, test this please.

@girving yes, on it!

**Comment by yanchen036** [comment]
```
ERROR: /workspace/tensorflow/stream_executor/BUILD:39:1: undeclared inclusion(s) in rule '//tensorflow/stream_executor:cuda_platform':
this rule is missing dependency declarations for the following files included by 'tensorflow/stream_executor/cuda/cuda_event.cc':
  'bazel-out/local_linux-opt/genfiles/external/local_config_cuda/cuda/include/cuda.h'.
```
@drpngx seems the error is about latest changes in 'tensorflow/stream_executor/cuda/cuda_event.cc'. I update my code to master, Could you test it again? Thank you.

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/slice_op_cpu_impl_6.cc (hunk 1 above)
Comment (drpngx) [inline tensorflow/core/kernels/slice_op_cpu_impl_6.cc]: The static part can be compiled into multiple files. Each branch of the switch statement goes into a separate file.

Comment (yanchen036) [inline tensorflow/core/kernels/slice_op_cpu_impl_6.cc]: Do you means these?
file: tensorflow/core/kernels/slice_op.h
```C++
template <typename Device, typename T, int NDIMS>
struct Slice {
  void operator()(const Device& d, Tensor* out, const Tensor& in,
                  const gtl::ArraySlice<int64>& slice_indices,
                  const gtl::ArraySlice<int64>& slice_sizes) {
    switch (in.dims()) {
      case NDIMS:
        internal::SliceUsingEigen<Device, T, NDIMS>(d, out, in, slice_indices, slice_sizes);
        break;
      default:
        if (Eigen::internal::is_same<Device, Eigen::GpuDevice>::value) {
          internal::SliceSimpleGpu<Device, T>(d, out, in, slice_indices);
        } else {
          internal::SliceSimple<Device, T>(d, out, in, slice_indices);
        }
        break;
    }
  }
};
```
and file: tensorflow/core/kernels/slice_op_cpu_impl_1.cc
```C++
#define CPU_PROVIDED_IXDIM 1
#include "tensorflow/core/kernels/slice_op_cpu_impl.h"
#undef CPU_PROVIDED_IXDIM
```

Comment (drpngx) [inline tensorflow/core/kernels/slice_op_cpu_impl_6.cc]: Yes something like that.

**Comment by yanchen036** [comment]
@drpngx add impl_*.cc files back, pls review and run test

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/slice_op.cc [hunk 2]
```diff
@@ -192,39 +192,23 @@ class SliceOp : public OpKernel {
...
     return;                                         \
   }
-
       HANDLE_DIM(1);
       HANDLE_DIM(2);
       HANDLE_DIM(3);
       HANDLE_DIM(4);
       HANDLE_DIM(5);
       HANDLE_DIM(6);
-      HANDLE_DIM(7);
```
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.cc]: Why not 7?

File: tensorflow/core/kernels/slice_op.cc [hunk 3]
```diff
@@ -264,24 +248,13 @@ class MklSliceOp : public OpKernel {
...
-      HANDLE_DIM(4);
-      HANDLE_DIM(5);
-      HANDLE_DIM(6);
-      HANDLE_DIM(7);
-
-#undef HANDLE_DIM
-
-      OP_REQUIRES(context, false, errors::Unimplemented(
-                                      "SliceOp : Unhandled input dimensions"));
+      // special case for handling 4-D tensor slice
```
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.cc]: "Special case for handling 4-d tensor slice."

File: tensorflow/core/kernels/slice_op.h [hunk 4]
```diff
@@ -19,31 +19,84 @@ limitations under the License.
...
+template <typename Device, typename T>
+void SliceSimpleGpu(const Device& d, Tensor* out, const Tensor& in,
+                 const gtl::ArraySlice<int64>& slice_indices);
+
+template <typename Device, typename T>
+void SliceSimple(const Device& d, Tensor* out, const Tensor& in,
+                 const gtl::ArraySlice<int64>& slice_indices) {
+  const int ndims = in.dims();
+  const int64 nelem = out->NumElements();
+  gtl::InlinedVector<int64, 8> in_strides = ComputeStride<int64>(in.shape());
```
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.h]: const

File: tensorflow/core/kernels/slice_op.h [hunk 5]
```diff
@@ -19,31 +19,84 @@ limitations under the License.
...
+  gtl::InlinedVector<int64, 8> out_strides = ComputeStride<int64>(out->shape());
```
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.h]: const

File: tensorflow/core/kernels/slice_op.h [hunk 6]
```diff
@@ -19,31 +19,84 @@ limitations under the License.
...
+  auto output = out->tensor<T, NDIMS>();
+  Eigen::DSizes<int, NDIMS> indices;
+  for (int i = 0; i < NDIMS; ++i) {
+    indices[i] = slice_indices[i];
+  }
+  Eigen::DSizes<int, NDIMS> sizes;
+  for (int i = 0; i < NDIMS; ++i) {
+    sizes[i] = slice_sizes[i];
+  }
+  bool use_64bit = (input.size() > Eigen::NumTraits<int>::highest());
```
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.h]: const. No need for parens.

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/slice_op.cc (hunk 2 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.cc]: case 7 is in line 208. It handle all the cases which dim >= 7

File: tensorflow/core/kernels/slice_op.cc (hunk 3 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.cc]: handled in line 304

File: tensorflow/core/kernels/slice_op.h (hunk 4 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.h]: ok

File: tensorflow/core/kernels/slice_op.h (hunk 5 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.h]: ok

File: tensorflow/core/kernels/slice_op.h (hunk 6 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.h]: ok

**Comment by yanchen036** [comment]
Is there anything I need to do?

File: tensorflow/core/kernels/slice_op.h [hunk 7]
```diff
@@ -19,31 +19,84 @@ limitations under the License.
...
+void SliceSimple(const Device& d, Tensor* out, const Tensor& in,
+                 const gtl::ArraySlice<int64>& slice_indices) {
+  const int ndims = in.dims();
+  const int64 nelem = out->NumElements();
+  const gtl::InlinedVector<int64, 8> in_strides = ComputeStride<int64>(in.shape());
+  const gtl::InlinedVector<int64, 8> out_strides = ComputeStride<int64>(out->shape());
+  const T* p = in.flat<T>().data();
+  T* q = out->flat<T>().data();
+
+  for (int64 o_idx = 0; o_idx < nelem; ++o_idx) {
```
Comment (aselle) [inline tensorflow/core/kernels/slice_op.h]: This implementation is correct but relatively slow. It doesn't use threading and uses a lot of divides and mods. I would add  TODO to indicate that this could be improved if performance of higher order tensor becomes a problem.

Comment (aselle) [inline tensorflow/core/kernels/slice_op.h]: Specifically, I think if you incremented the input index in non-collapsed form it might be somewhat faster.

**Review by aselle — COMMENTED** [review]
LGTM generally, but I would add a benchmark for this in the vein of the one I wrote for strided_slice_op_test.cc 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/strided_slice_op_test.cc

It would be interesting to know how much slower the higher dimension version is than the main version i.e. try two slices one with say a d slice and one with an 8d slice.

**Comment by yanchen036** [comment]
@aselle I've added some benchmark, Please take a look.

File: tensorflow/core/kernels/slice_op.h (hunk 7 above)
Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.h]: Yes, you're right, but I've no idea how to uncollapse the for loop. Do you have any suggestion?

**Comment by rmlarsen** [comment]
@aselle @drpngx can you please take another look?

**Comment by jhseu** [comment]
@aselle @drpngx friendly ping

File: tensorflow/core/kernels/slice_op.cc (hunk 3 above)
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.cc]: I mean, please capitalize and punctuate comment.

Comment (yanchen036) [inline tensorflow/core/kernels/slice_op.cc]: Done.

File: tensorflow/core/kernels/slice_op.h (hunk 7 above)
Comment (drpngx) [inline tensorflow/core/kernels/slice_op.h]: You can always unroll loops with Duff's device which batches up the divisions by 8, not if that's what Andy meant. It feels like the inner loop has lots of computation repeated for every element, so if you can work out a version where you have ndims on the outer and nelems on the inner loop then it would be faster.

It would be nice to measure the improvement on a benchmark.

Comment (yanchen036) [reply tensorflow/core/kernels/slice_op.h]: How about use a buffer to record all outer's i_idx and t, in this way, I can switch the inner loop and outer loop. It's space-time tradeoff.

Comment (drpngx) [reply tensorflow/core/kernels/slice_op.h]: Sounds like a reasonable trade.

**Comment by martinwicke** [comment]
@aselle @drpngx ping?

**Comment by martinwicke** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/slice_op.h (hunk 7 above)
Comment (yanchen036) [reply tensorflow/core/kernels/slice_op.h]: @drpngx uncollapse the inner loop, pls review it.

**Comment by yanchen036** [comment]
@drpngx uncollapse the for loop by Duff's device, pls take a look. Thank you.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
If you had some benchmarks for before/after, please report them here. That would help.

**Comment by yanchen036** [comment]
Yes, but I can't find the output of benchmark result. I read the rest log, there isn't benchmark result there. Could you tell me where to get the benchmark result? thanks.

**Comment by sb2nov** [comment]
@drpngx ping

**Review by drpngx — APPROVED** [review]

**Comment by drpngx** [comment]
We don't have a benchmark for slice. Can you create a program to run on a large number of random values, exercising both cases when the outer and inner loop have different sizes?

**Comment by sb2nov** [comment]
Jenkins, test this please.

**Comment by yanchen036** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
Could you rebase this PR on top of master?
Also, was the comment by @drpngx addressed?

File: tensorflow/python/kernel_tests/slice_op_test.py [hunk 8]
```diff
@@ -227,6 +251,7 @@ def testPartialShapeInference(self):
     m2 = array_ops.slice(z, [0, 0, 0], [constant_op.constant(1) + 0, 2, -1])
     self.assertAllEqual(m2.get_shape().as_list(), [None, 2, None])
 
+>>>>>>> upstream/master
```
Comment (gunan) [inline tensorflow/python/kernel_tests/slice_op_test.py]: Forgotten from the merge?

**Comment by yanchen036** [comment]
sorry, removed line ">>>>>>>"

**Comment by gunan** [comment]
Jenkins, test this please.
